File: Python/isBalanced.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: isBalanced.py
@Function: 平衡二叉树 
@Link: https://leetcode-cn.com/problems/balanced-binary-tree/
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-07-31
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        def height(root: TreeNode) -> int:
            if not root:
                return 0
            leftHeight = height(root.left)
            rightHeight = height(root.right)
            if leftHeight == -1 or rightHeight == -1 or abs(leftHeight - rightHeight) > 1:
                return -1
            else:
                return max(leftHeight, rightHeight) + 1

        return height(root) >= 0


# ---------------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass

File: C++/isBalanced.cpp
/**
 * @File    : isBalanced.cpp
 * @Brief   : 平衡二叉树
 * @Link    : https://leetcode-cn.com/problems/balanced-binary-tree/
 * @Author  : Wei Li
 * @Date    : 2021-07-31
*/

#include <iostream>
#include <algorithm>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    // 主函数
    bool isBalanced(TreeNode *root)
    {
        return helper(root) != -1;
    }
    // 辅函数
    int helper(TreeNode *root)
    {
        if (!root)
        {
            return 0;
        }
        int left = helper(root->left), right = helper(root->right);
        if (left == -1 || right == -1 || abs(left - right) > 1)
        {
            return -1;
        }
        return 1 + std::max(left, right);
    }
};

// ------------------------------
int main(int argc, char **argv)
{
    // test on LeetCode online.

    return 0;
}
